File: src/core/undo/SimpleUndo.ts
import { Undo } from "./Undo";
import { Action } from "./Action";
import { Property } from "../observable/property/Property";
import { map, property } from "../observable";
import { undo_manager } from "./UndoManager";
import { WritableProperty } from "../observable/property/WritableProperty";

/**
 * Simply contains a single action. `can_undo` and `can_redo` must be managed manually.
 */
export class SimpleUndo implements Undo {
    readonly action: WritableProperty<Action>;

    constructor(description: string, undo: () => void, redo: () => void) {
        this.action = property({ description, undo, redo });

        this.first_undo = map(
            (action, can_undo) => (can_undo ? action : undefined),
            this.action,
            this.can_undo,
        );

        this.first_redo = map(
            (action, can_redo) => (can_redo ? action : undefined),
            this.action,
            this.can_redo,
        );
    }

    make_current(): void {
        undo_manager.current.val = this;
    }

    readonly can_undo = property(false);

    readonly can_redo = property(false);

    readonly first_undo: Property<Action | undefined>;

    readonly first_redo: Property<Action | undefined>;

    undo(): boolean {
        if (this.can_undo) {
            this.action.val.undo();
            return true;
        } else {
            return false;
        }
    }

    redo(): boolean {
        if (this.can_redo) {
            this.action.val.redo();
            return true;
        } else {
            return false;
        }
    }

    reset(): void {
        this.can_undo.val = false;
        this.can_redo.val = false;
    }
}

File: web/src/main/kotlin/world/phantasmal/web/core/undo/SimpleUndo.kt
package world.phantasmal.web.core.undo

import world.phantasmal.observable.value.*
import world.phantasmal.web.core.actions.Action

/**
 * Simply contains a single action. [canUndo] and [canRedo] must be managed manually.
 */
class SimpleUndo(
    undoManager: UndoManager,
    private val description: String,
    undo: () -> Unit,
    redo: () -> Unit,
) : Undo {
    private val action = object : Action {
        override val description: String = this@SimpleUndo.description

        override fun execute() {
            redo()
        }

        override fun undo() {
            undo()
        }
    }

    override val canUndo: MutableVal<Boolean> = mutableVal(false)
    override val canRedo: MutableVal<Boolean> = mutableVal(false)

    override val firstUndo: Val<Action?> = canUndo.map { if (it) action else null }
    override val firstRedo: Val<Action?> = canRedo.map { if (it) action else null }

    init {
        undoManager.addUndo(this)
    }

    override fun undo(): Boolean =
        if (canUndo.value) {
            action.undo()
            true
        } else {
            false
        }

    override fun redo(): Boolean =
        if (canRedo.value) {
            action.execute()
            true
        } else {
            false
        }

    override fun reset() {
        canUndo.value = false
        canRedo.value = false
    }
}
